민초/04-콜백함수.md

```diff
@@ -0,0 +1,337 @@
+  ```
+
+- Promise Chaining시 에러 처리
+
+  ```js
+  promiseGet(`${url}/posts/1`)
+    .then(({ userId }) => promiseGet(`${url}/users/${userId}`))
+    .then(userInfo => console.log(userInfo))
+    .catch(err => console.error(err));
+  ```
+
+### Promise all, race, settled
+
+- **Promise.all** : 여러 개의 비동기 처리를 모두 병렬 처리할 때 사용.
+
+  ```js
+  const requestData1 = () =>
+  	new Promise(resolve => setTimeout(() => resolve(1), 3000));
+  const requestData2 = () =>
+  	new Promise(resolve => setTimeout(() => resolve(2), 2000));
+  const requestData3 = () =>
+  	new Promise(resolve => setTimeout(() => resolve(3), 1000));
+  
+  Promise.all([requestData1(), requestData2(), requestData3()])
+  	.then(console.log) // [ 1, 2, 3] 
+  	.catch(console.error);
+  ```
+
+  - Promise.all은 프로미스를 요소로 갖는 배열 등의 이터러블을 인수로 전달 받고, 전달 받은 모든 프로미스가 모두 fulfilled 상태가 되면 모든 처리 결과를 배열에 저장해 새로운 프로미스를 반환.
+  - fulfilled 상태가 되는 순서와 상관없이 배열에 저장함. 즉, 처리 순서가 보장.
+  - Promise.all은 인수로 전달받은 배열의 프로미스가 하나라도 rejected 상태가 되면, 나머지 프로미스가 fulfilled 상태가 되는 것을 기다리지 않고 즉시 종료함.
+
+- **Promise.race**
+
+  - 가장 먼저 fulfilled 상태가 된 프로미스의 처리 결과를 resolve하는 새로운 프로미스를 반환.
+
+    ```js
+    Promise.race([
+        new Promise(resolve => setTimeout(() => resolve(1), 3000)),
+        new Promise(resolve => setTimeout(() => resolve(2), 2000)),
+        new Promise(resolve => setTimeout(() => resolve(3), 1000))
+    ])
+      .then(console.log) // 3
+      .catch(console.log);
+    ```
+
+  - Promise.all과 마찬가지로 인수로 전달받은 배열의 프로미스가 하나라도 rejected 상태가 되면 에러를 reject하는 새로운 프로미스를 즉시 반환.
+
+    ```js
+    Promise.race([
+        new Promise((_, reject) => setTimeout(() => reject(new Error('Error 1'), 3000)),
+        new Promise((_, reject) => setTimeout(() => reject(new Error('Error 2'), 2000)),
+        new Promise((_, reject) => setTimeout(() => reject(new Error('Error 3'), 1000))
+    ])
+      .then(console.log)
+      .catch(console.log); // Error: Error 3
+    ```
+
+- **Promise.allSettled**
+
+  - 전달받은 프로미스가 모두 settled 상태가 되면 처리 결과를 배열로 반환.
+
+    ```js
+    Promise.allSettled([
+        new Promise(resolve => setTimeout(() => resolve(1), 2000)),
+        new Promise((_, reject) => setTimeOut(() => reject(new Error('Error!')), 1000))
+    ]).then(console.log);
+    
+    /*
+    [
+      { status: 'fulfilled', value: 1 },
+      {
+        status: 'rejected',
+        reason: Error: 'setTimeOut is not defined'
+      }
+    ]
+    */
+    ```
+
+### 마이크로태스크 큐?
+
+- 어떤 순서로 로그가 출력될까?
+
+  ```js
+  setTimeout(() => console.log(1), 0);
+  
+  Promise.resolve()
+  	.then(() => console.log(2))
+  	.then(() => console.log(3));
+  ```
+
+- 1 -> 2 -> 3? 응~ 아니야
+
+- 2 -> 3 -> 1로 출력됨. 왜??
+
+- 이렇게 실행되는 이유는 프로미스의 후속 처리 메서드의 콜백함수는 태스크 큐가 아니라 `마이크로태스크 큐`에 저장되기 때문.
+
+- `마이크로태스크 큐`는 태스크 큐와는 별도의 큐로, 프로미스의 후속 처리 메서드의 콜백 함수가 일시 저장됨.
+
+- 그 외 비동기 함수의 콜백 함수나 이벤트 핸들러는 `태스크 큐`에 일시 저장됨.
+
+- **마이크로태스크 큐는 태스크 큐보다 우선순위가 높음.** 
+
+- 즉, 이벤트 루프는 콜 스택이 비면 마이크로태스크 큐에 대기하고 있는 함수를 먼저 가져와 실행하고, 마이크로태스크 큐가 비면 태스크 큐에 대기하고 있는 함수를 가져와 실행.
```

Comment:
마이크로태스크 큐와 태스크 큐는 큐 내에 있는 함수를 가져올 때 그냥 단순히 가장 첫 번째에 있는 함수를 가져오나요? 아니면 서로 차이가 있나요?

Reply:
기본적으로는 FIFO이죠..?
사실 제가 새벽에 마이크로태스크 큐와 태스크 큐의 차이점을 검색해보다가 [이런 글](https://velog.io/@stseo012/JavaScript#task-queuemacro-queue%EC%99%80-microtask-queue)을 보게 되었는데요. 여기에서는 `Task queues are sets, not queues, because step one of the event loop processing model grabs the first runnable task from the chosen queue, instead of dequeuing the first task.`, 즉 `Task Queue는 Task의 집합이다. 단순히 큐의 첫 번째 Task를 가져오는 것이 아니라, 실행가능한 가장 오래된 Task를 가져온다.`라고 표현을 해놨더라고요. 그래서 엄밀히 말해 둘이 꺼내는 방식에 차이가 있는 것인지가 궁금했어요~!